Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp

```diff
@@ -0,0 +1,505 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "MemoryMappedGPUBuffer.h"
+
+#if USE(GBM)
+#include "DRMDeviceManager.h"
+#include "DRMDeviceNode.h"
+#include "GBMVersioning.h"
+#include "GLDisplay.h"
+#include "Logging.h"
+#include "PlatformDisplay.h"
+
+#include <epoxy/egl.h>
+#include <linux/dma-buf.h>
+#include <sys/ioctl.h>
+#include <wtf/SafeStrerror.h>
+#include <wtf/unix/UnixFileDescriptor.h>
+
+#if USE(LIBDRM)
+#include <drm_fourcc.h>
+#endif
+
+namespace WebCore {
+
+constexpr bool dumpSupportedDRMFormatsAndModifiers = false;
+constexpr bool dumpBufferFormatNegotiation = false;
+constexpr bool dumpBufferAllocation = false;
+constexpr bool useDRMModifiers = true;
+
+MemoryMappedGPUBuffer::MemoryMappedGPUBuffer(const IntSize& size)
+    : m_size(size)
+{
+}
+
+MemoryMappedGPUBuffer::~MemoryMappedGPUBuffer()
+{
+    if (m_mappedData)
+        unmap();
+
+    if (m_dmaBufFD > 0)
+        close(m_dmaBufFD);
+
+    if (m_bo)
+        gbm_bo_destroy(m_bo);
+
+    auto& display = WebCore::PlatformDisplay::sharedDisplay();
+    display.destroyEGLImage(m_eglImage);
+}
+
+static bool shouldPreferDRMLinearModifier()
+{
+    static std::once_flag onceFlag;
+    static bool shouldPreferLinear = false;
+
+    std::call_once(onceFlag, [] {
+        if (const char* envString = getenv("WEBKIT_GPU_BUFFERS_PREFER_LINEAR_MODIFIER")) {
+            auto envStringView = StringView::fromLatin1(envString);
+            if (envStringView == "1"_s)
+                shouldPreferLinear = true;
+        }
+    });
+
+    return shouldPreferLinear;
+}
+
+std::unique_ptr<MemoryMappedGPUBuffer> MemoryMappedGPUBuffer::create(const IntSize& size, bool isOpaque)
+{
+    auto& manager = WebCore::DRMDeviceManager::singleton();
+    ASSERT(manager.isInitialized());
+
+    auto deviceNode = manager.mainDeviceNode(WebCore::DRMDeviceManager::NodeType::Primary);
+    if (deviceNode)
+        LOG(Images, "[MemoryMappedGPUBuffer] Allocating dma-bufs from device '%s'", deviceNode->filename().data());
```

Comment:
```suggestion
    if (!deviceNode) {
        LOG(Images, "[MemoryMappedGPUBuffer] Failed to get device node");
        return nullptr;
    }

    LOG(Images, "[MemoryMappedGPUBuffer] Allocating dma-bufs from device '%s'", deviceNode->filename().data());
```

We should have a static checker that flags the above.

If we consider `deviceNode` worth a null check before dereferencing it to log, then we should not unconditionally dereference it to get the `gbmDevice()`

Reply:
Good catch, will adapt to:

```
diff --git a/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp b/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp
index 6dde788c162c..0e3045818331 100644
--- a/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp
+++ b/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp
@@ -71,8 +71,12 @@ std::unique_ptr<MemoryMappedGPUBuffer> MemoryMappedGPUBuffer::create(const IntSi
     ASSERT(manager.isInitialized());
 
     auto deviceNode = manager.mainDeviceNode(WebCore::DRMDeviceManager::NodeType::Render);
-    if (deviceNode)
-        LOG(GPUBuffers, "MemoryMappedGPUBuffer::create(), allocating dma-bufs from device '%s'", deviceNode->filename().data());
+    if (!deviceNode) {
+        LOG(GPUBuffers, "MemoryMappedGPUBuffer::create(), failed to get GBM render device node");
+        return nullptr;
+    }
+
+    LOG(GPUBuffers, "MemoryMappedGPUBuffer::create(), allocating dma-bufs from device '%s'", deviceNode->filename().data());
 
     auto* device = deviceNode->gbmDevice();
     if (!device) {
```